src/mpid/ch4/include/mpidch4.h

```diff
@@ -418,6 +418,7 @@ int MPID_Waitall_enqueue(int count, MPI_Request * array_of_requests,
                          MPI_Status * array_of_statuses);
 int MPID_Abort(struct MPIR_Comm *comm, int mpi_errno, int exit_code, const char *error_msg);
 
+int MPID_Pset_member_get_lpid(MPIR_Pset_member * member_ptr, uint64_t * lpid_ptr);
```

Comment:
On a second thought, this may not be sufficient. The idea is for the device layer to establish connections or perform address exchanges. But it can't be done with a local function such as this. So I am thinking that it has to be a collective call. For example, `MPID_Group_from_pset` - to be called from `MPIR_Group_from_session_pset_impl`.

I am giving it a stab. Will update.

Reply:
> So I am thinking that it has to be a collective call

One thing to keep in mind is that "MPI_GROUP_FROM_SESSION_PSET is a local function" according to the MPI standard. So we cannot actually do real collective calls here.

Reply:
Right :( 

That only leaves one option -- to refactor the current `MPIR_Group` to be built on `device-independent`  "pset member id", and delay the necessary connection/address exchange until `MPI_Comm_create_group`.

That also simplifies the design since there won't be need for *device-dependent* lpid. The device layer just need manage address lookup from *device-independent* lpid. Unfortunately, that will be substantial changes. I'll give it a try.